README.md
<p align="center"><img src="screenshots/reviewer.gif"></p>

<h2 align="center">Visual Feedback for Reviews for Anki</h2>

<p align="center">
<a title="Latest (pre-)release" href="https://github.com/glutanimate/review-feedback/releases"><img src ="https://img.shields.io/github/release-pre/glutanimate/review-feedback.svg?colorB=brightgreen"></a>
<a title="License: GNU AGPLv3" href="https://github.com/glutanimate/review-feedback/blob/master/LICENSE"><img  src="https://img.shields.io/badge/license-GNU AGPLv3-green.svg"></a>
<a title="Rate on AnkiWeb" href="https://ankiweb.net/shared/info/1749604199"><img src="https://glutanimate.com/logos/ankiweb-rate.svg"></a>
<br>
<a title="Buy me a coffee :)" href="https://ko-fi.com/X8X0L4YV"><img src="https://img.shields.io/badge/ko--fi-contribute-%23579ebd.svg"></a>
<a title="Support me on Patreon :D" href="https://www.patreon.com/bePatron?u=7522179"><img src="https://img.shields.io/badge/patreon-support-%23f96854.svg"></a>
<a title="Follow me on Twitter" href="https://twitter.com/intent/user?screen_name=glutanimate"><img src="https://img.shields.io/twitter/follow/glutanimate.svg"></a>
</p>

> Positive reinforcement for your Anki reviews! 

This is an add-on for the spaced-repetition flashcard app [Anki](https://apps.ankiweb.net/). It provides users with visual **feedback** for reviews by **flashing** a **small transparent image** at the center of the screen that varies between lapses and passed cards.

### Table of Contents <!-- omit in toc -->

<!-- MarkdownTOC levels="1,2,3" -->

- [Installation](#installation)
- [Documentation](#documentation)
- [Building](#building)
- [Contributing](#contributing)
- [License and Credits](#license-and-credits)

<!-- /MarkdownTOC -->

<!-- ### Screenshots

![](screenshots/screenshot.png) -->

### Installation

#### AnkiWeb <!-- omit in toc -->

The easiest way to install Visual Feedback for Reviews is through [AnkiWeb](https://ankiweb.net/shared/info/1749604199).

#### Manual installation <!-- omit in toc -->

Please click on the entry corresponding to your Anki version:

<details>

<summary><i>Anki 2.1</i></summary>

1. Make sure you have the [latest version](https://apps.ankiweb.net/#download) of Anki 2.1 installed. Earlier releases (e.g. found in various Linux distros) do not support `.ankiaddon` packages.
2. Download the latest `.ankiaddon` package from the [releases tab](https://github.com/glutanimate/review-feedback/releases) (you might need to click on *Assets* below the description to reveal the download links)
3. From Anki's main window, head to *Tools* → *Add-ons*
4. Drag-and-drop the `.ankiaddon` package onto the add-ons list
5. Restart Anki

Video summary:

<img src="https://raw.githubusercontent.com/glutanimate/docs/master/anki/add-ons/media/ankiaddon-installation.gif" width=640>

</details>

### Documentation

For further information on the use of this add-on please check out [the description text](docs/description.md) for AnkiWeb.

### Building

With [Anki add-on builder](https://github.com/glutanimate/anki-addon-builder/) installed:

    git clone https://github.com/glutanimate/review-feedback.git
    cd review-feedback
    aab build

For more information on the build process please refer to [`aab`'s documentation](https://github.com/glutanimate/anki-addon-builder/#usage).

### Contributing

Contributions are welcome! Please review the [contribution guidelines](./CONTRIBUTING.md) on how to:

- Report issues
- File pull requests
- Support the project as a non-developer

### License and Credits

*Visual Feedback for Reviews* is *Copyright © 2017-2020 [Aristotelis P.](https://glutanimate.com/) (Glutanimate)*

Visual Feedback for Reviews is free and open-source software. The add-on code that runs within Anki is released under the GNU AGPLv3 license, extended by a number of additional terms. For more information please see the [LICENSE](https://github.com/glutanimate/review-feedback/blob/master/LICENSE) file that accompanied this program.

Please note that this program uses the [Libaddon](https://github.com/glutanimate/anki-libaddon/) library which comes with [its own additional terms extending the GNU AGPLv3 license](https://github.com/glutanimate/anki-libaddon/blob/master/LICENSE). You may only copy, distribute, or modify the present compilation of this program with Libaddon under the combined licensing terms specified by both licenses.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY.

----

<b>
<div align="center">The continued development of this add-on is made possible <br>thanks to my <a href="https://www.patreon.com/glutanimate">Patreon</a> and <a href="https://ko-fi.com/X8X0L4YV">Ko-Fi</a> supporters.
<br>You guys rock ❤️ !</div>
</b>

CHANGELOG.md
# Changelog

All notable changes to Visual Feedback for Reviews will be documented here. You can click on each release number to be directed to a detailed log of all code commits for that particular release. The download links will direct you to the GitHub release page, allowing you to manually install a release if you want.

If you enjoy Visual Feedback for Reviews, please consider supporting my work on Patreon, or by buying me a cup of coffee :coffee::

<p align="center">
<a href="https://www.patreon.com/glutanimate" rel="nofollow" title="Support me on Patreon 😄"><img src="https://glutanimate.com/logos/patreon_button.svg"></a>      <a href="https://ko-fi.com/X8X0L4YV" rel="nofollow" title="Buy me a coffee 😊"><img src="https://glutanimate.com/logos/kofi_button.svg"></a>
</p>

:heart: My heartfelt thanks goes out to everyone who has supported this add-on through their tips, contributions, or any other means (you know who you are!). All of this would not have been possible without you. Thank you for being awesome!

## [Unreleased]

## [1.0.0] - 2020-03-04

### [Download](https://github.com/glutanimate/review-feedback/releases/tag/v1.0.0)

### Added

![](https://raw.githubusercontent.com/glutanimate/review-feedback/master/screenshots/configuration.gif)

- Support for Anki 2.1
- Configuration via Anki's built-in config dialog
- New `imageSet` option, providing the ability to quickly switch between different image sets
- Support for installing additional image sets via the `user_files` directory

### Fixed

- Fixed off-center image position (thanks to @zavarka for the report)

### Changed

- Dropped Anki 2.0 support
- Mostly rewrote the add-on

## [1.0.0-beta.3] - 2020-01-29

### [Download](https://github.com/glutanimate/review-feedback/releases/tag/v1.0.0-beta.3)

### Fixed

- Fixed black image background on Windows
- Fixed support for Anki 2.1.20 beta6

### Changed

- Removed image drop shadow on macOS to bring the visuals in line with other platforms

## [1.0.0-beta.2] - 2020-01-28

### [Download](https://github.com/glutanimate/review-feedback/releases/tag/v1.0.0-beta.2)

### Fixed

- Fixed incompatibility with other add-ons using libaddon (typing-related vendoring troubles on Anki <2.1.16)

## [1.0.0-beta.1] - 2020-01-23

### [Download](https://github.com/glutanimate/review-feedback/releases/tag/v1.0.0-beta.1)

### Added

- Support for Anki 2.1
- Configuration via Anki's built-in config dialog
- New `imageSet` option, providing the ability to quickly switch between different image sets
- Support for installing additional image sets via the `user_files` directory

### Fixed

- Fixed off-center image position (thanks to @zavarka for the report)

### Changed

- Dropped Anki 2.0 support
- Mostly rewrote the add-on

## [v0.1.1] - 2017-08-06

### Added

- Basic configuration capabilities

## v0.1.0 - 2017-06-15

### Added

- Initial release of Visual Feedback for Reviews, forked from an earlier add-on by nest0r/Ja-Dark

[Unreleased]: https://github.com/glutanimate/review-feedback/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/glutanimate/review-feedback/compare/v1.0.0-beta.3...v1.0.0
[1.0.0-beta.3]: https://github.com/glutanimate/review-feedback/compare/v1.0.0-beta.2...v1.0.0-beta.3
[1.0.0-beta.2]: https://github.com/glutanimate/review-feedback/compare/v1.0.0-beta.1...v1.0.0-beta.2
[1.0.0-beta.1]: https://github.com/glutanimate/review-feedback/compare/v0.1.1...v1.0.0-beta.1
[0.1.1]: https://github.com/glutanimate/anjoy/releases/tag/v0.1.1

-----

The format of this file is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

src/review_feedback/config.py
# -*- coding: utf-8 -*-

# Visual Feedback for Reviews Add-on for Anki
#
# Copyright (C) 2017-2020  Aristotelis P. <https://glutanimate.com/>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version, with the additions
# listed at the end of the license file that accompanied this program.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# NOTE: This program is subject to certain additional terms pursuant to
# Section 7 of the GNU Affero General Public License.  You should have
# received a copy of these additional terms immediately following the
# terms and conditions of the GNU Affero General Public License that
# accompanied this program.
#
# If not, please request a copy through one of the means of contact
# listed here: <https://glutanimate.com/contact/>.
#
# Any modifications to this file must keep this entire header intact.

"""
Handles add-on configuration
"""

from aqt import mw

from .libaddon.anki.configmanager import ConfigManager

config = ConfigManager(mw)


src/review_feedback/tooltip.py
# -*- coding: utf-8 -*-

# Visual Feedback for Reviews Add-on for Anki
#
# Copyright (C) 2017-2020  Aristotelis P. <https://glutanimate.com/>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version, with the additions
# listed at the end of the license file that accompanied this program.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# NOTE: This program is subject to certain additional terms pursuant to
# Section 7 of the GNU Affero General Public License.  You should have
# received a copy of these additional terms immediately following the
# terms and conditions of the GNU Affero General Public License that
# accompanied this program.
#
# If not, please request a copy through one of the means of contact
# listed here: <https://glutanimate.com/contact/>.
#
# Any modifications to this file must keep this entire header intact.

"""
Module template.
"""

def initializeTooltip():
    from aqt.utils import tooltip
    tooltip("Visual Feedback for Reviews running!")


src/review_feedback/__init__.py
# -*- coding: utf-8 -*-

# Visual Feedback for Reviews Add-on for Anki
#
# Copyright (C) 2017-2020  Aristotelis P. <https://glutanimate.com/>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version, with the additions
# listed at the end of the license file that accompanied this program.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# NOTE: This program is subject to certain additional terms pursuant to
# Section 7 of the GNU Affero General Public License.  You should have
# received a copy of these additional terms immediately following the
# terms and conditions of the GNU Affero General Public License that
# accompanied this program.
#
# If not, please request a copy through one of the means of contact
# listed here: <https://glutanimate.com/contact/>.
#
# Any modifications to this file must keep this entire header intact.

"""
Initializes add-on components.
"""

try:
    import typing  # noqa: F401
    import types  # noqa: F401
except ImportError:
    from .libaddon import registerLegacyVendorDir

    registerLegacyVendorDir()


from ._version import __version__  # noqa: F401


def checkFor2114ImportError():
    from .consts import ADDON

    try:
        # litmus test for Anki import bug
        from .libaddon.platform import anki_version  # type: ignore

        return True
    except ImportError:
        # Disable add-on and inform user of the bug
        from aqt.utils import showWarning
        from aqt import mw
        from anki import version as anki_version

        mw.addonManager.toggleEnabled(__name__, enable=False)

        bug = "https://anki.tenderapp.com/discussions/ankidesktop/34836"
        downloads = "https://apps.ankiweb.net#download"
        beta = "https://apps.ankiweb.net/downloads/beta/"
        vers = "2.1.15"
        title = "Warning: {name} disabled".format(name=ADDON.NAME)
        msg = (
            "<b>WARNING</b>: {name} had to be disabled because the "
            "version of Anki that is currently installed on your system "
            "({anki_version}) is incompatible with the add-on.<br><br> "
            "Earlier releases of Anki like this one "
            "suffer from a <a href='{bug}'>bug</a> that breaks "
            "{name} and many other add-ons on your system. "
            "In order to fix this you will have to update Anki "
            "to version {vers} or higher.<br><br>"
            "As of writing this message, Anki {vers} is still in "
            "beta testing, but that might have "
            "changed in the meantime. Please check with the "
            "<a href='{downloads}'>releases page</a> to see if {vers} "
            "or a later release is available, otherwise download and "
            "install the 2.1.15 beta <a href='{beta}'>here</a>.<br><br>"
            "After updating Anki, please re-enable "
            "{name} by heading to Tools → Add-ons, selecting the "
            "add-on, and clicking <i>Toggle Enabled</i>.".format(
                name=ADDON.NAME,
                anki_version=anki_version,
                bug=bug,
                vers=vers,
                downloads=downloads,
                beta=beta,
            )
        )

        showWarning(msg, title=title, textFormat="rich")

        return False


def initializeAddon():
    """Initializes add-on after performing a few checks
    
    Allows more fine-grained control over add-on execution, which can
    be helpful when implementing workarounds for Anki bugs (e.g. the module
    import bug present in all Anki 2.1 versions up to 2.1.14)
    """

    if not checkFor2114ImportError():
        return False

    from .consts import ADDON
    from .libaddon.consts import setAddonProperties

    setAddonProperties(ADDON)

    # from .libaddon.debug import maybeStartDebugging

    # maybeStartDebugging()

    from .reviewer import initializeReviewer

    initializeReviewer()


initializeAddon()


src/review_feedback/_version.py
# -*- coding: utf-8 -*-

# Visual Feedback for Reviews Add-on for Anki
#
# Copyright (C) 2017-2020  Aristotelis P. <https://glutanimate.com/>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version, with the additions
# listed at the end of the license file that accompanied this program.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# NOTE: This program is subject to certain additional terms pursuant to
# Section 7 of the GNU Affero General Public License.  You should have
# received a copy of these additional terms immediately following the
# terms and conditions of the GNU Affero General Public License that
# accompanied this program.
#
# If not, please request a copy through one of the means of contact
# listed here: <https://glutanimate.com/contact/>.
#
# Any modifications to this file must keep this entire header intact.

"""
Version information
"""

__version__ = "1.0.0"


src/review_feedback/feedback.py
# -*- coding: utf-8 -*-

# Visual Feedback for Reviews Add-on for Anki
#
# Copyright (C) 2017-2020  Aristotelis P. <https://glutanimate.com/>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version, with the additions
# listed at the end of the license file that accompanied this program.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# NOTE: This program is subject to certain additional terms pursuant to
# Section 7 of the GNU Affero General Public License.  You should have
# received a copy of these additional terms immediately following the
# terms and conditions of the GNU Affero General Public License that
# accompanied this program.
#
# If not, please request a copy through one of the means of contact
# listed here: <https://glutanimate.com/contact/>.
#
# Any modifications to this file must keep this entire header intact.

"""
Visual feedback
"""

from PyQt5.QtGui import QPixmap
from PyQt5.QtWidgets import QLabel
from PyQt5.QtCore import QPoint, Qt, QTimer

from aqt import mw

from typing import Optional

_lab: Optional[QLabel] = None
_timer: Optional[QTimer] = None


def closeConfirm():
    global _lab, _timer
    if _lab:
        try:
            _lab.deleteLater()
        except:  # noqa: E722
            pass
        _lab = None
    if _timer:
        _timer.stop()
        _timer = None


def confirm(image_path: str, period: int):
    global _timer, _lab
    parent = mw
    closeConfirm()
    lab = QLabel(parent=parent)
    img = QPixmap(image_path)
    lab.setPixmap(img)
    lab.setAttribute(Qt.WA_TranslucentBackground, True)
    lab.setAutoFillBackground(False)
    lab.setWindowFlags(
        Qt.ToolTip | Qt.FramelessWindowHint | Qt.NoDropShadowWindowHint  # type: ignore
    )
    center = parent.frameGeometry().center()
    qp = QPoint(img.width() * 0.5, img.height() * 0.5)  # type: ignore
    lab.move(center - qp)
    lab.show()
    _timer = mw.progress.timer(period, closeConfirm, False)
    _lab = lab


src/review_feedback/reviewer.py
# -*- coding: utf-8 -*-

# Visual Feedback for Reviews Add-on for Anki
#
# Copyright (C) 2017-2020  Aristotelis P. <https://glutanimate.com/>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version, with the additions
# listed at the end of the license file that accompanied this program.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# NOTE: This program is subject to certain additional terms pursuant to
# Section 7 of the GNU Affero General Public License.  You should have
# received a copy of these additional terms immediately following the
# terms and conditions of the GNU Affero General Public License that
# accompanied this program.
#
# If not, please request a copy through one of the means of contact
# listed here: <https://glutanimate.com/contact/>.
#
# Any modifications to this file must keep this entire header intact.

from pathlib import Path

from anki.cards import Card
from anki.hooks import wrap
from aqt.reviewer import Reviewer
from aqt.utils import showWarning

from .libaddon.platform import pathUserFiles, PATH_THIS_ADDON
from .feedback import confirm
from .config import config
from .consts import ADDON

from typing import Optional, NamedTuple, Callable

_lapsed_name = "passed.png"
_passed_name = "lapsed.png"


class MediaPaths(NamedTuple):
    passed: str
    lapsed: str


def _getImagePaths(set_name: str) -> Optional[MediaPaths]:
    default_path = Path(PATH_THIS_ADDON) / "images"
    user_path = Path(pathUserFiles()) / "images"

    for path in (default_path, user_path):
        path_passed = path / set_name / _passed_name
        path_lapsed = path / set_name / _lapsed_name

        if path_passed.is_file() and path_lapsed.is_file():
            return MediaPaths(str(path_passed), str(path_lapsed))

    return None


def onAnswerCard(reviewer: Reviewer, card: Card, ease: int):
    image_set = config["local"]["imageSet"]
    duration = config["local"]["feedbackDuration"]

    image_paths = _getImagePaths(image_set)

    if image_paths is None:
        showWarning(
            f"{ADDON.NAME} is not configured correctly: Could not find images for "
            f"'{image_set}'' image set. Please reset the configuration to the defaults "
            "and try again. If that does not work, please redownload the add-on."
        )
        return

    if ease == 1:
        confirm(image_paths.passed, duration)
    elif ease in (2, 3, 4):
        confirm(image_paths.lapsed, duration)


def onAnswerCardWrapper(reviewer: Reviewer, ease: int, _old: Callable):
    """Legacy wrapper for Anki <2.1.20"""
    if reviewer.mw.state != "review" or reviewer.state != "answer":
        return
    ret = _old(reviewer, ease)
    onAnswerCard(reviewer, ease)
    return ret


def initializeReviewer():
    try:
        from aqt.gui_hooks import reviewer_did_answer_card

        reviewer_did_answer_card.append(onAnswerCard)
    except (ImportError, ModuleNotFoundError):  # Anki < 2.1.20
        Reviewer._answerCard = wrap(Reviewer._answerCard, onAnswerCardWrapper, "around")


src/review_feedback/consts.py
# -*- coding: utf-8 -*-

# Visual Feedback for Reviews Add-on for Anki
#
# Copyright (C) 2017-2020  Aristotelis P. <https://glutanimate.com/>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version, with the additions
# listed at the end of the license file that accompanied this program.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# NOTE: This program is subject to certain additional terms pursuant to
# Section 7 of the GNU Affero General Public License.  You should have
# received a copy of these additional terms immediately following the
# terms and conditions of the GNU Affero General Public License that
# accompanied this program.
#
# If not, please request a copy through one of the means of contact
# listed here: <https://glutanimate.com/contact/>.
#
# Any modifications to this file must keep this entire header intact.

"""
Addon-wide constants
"""

from ._version import __version__

try:
    from .data.patrons import MEMBERS_CREDITED, MEMBERS_TOP  # type: ignore
except ImportError:
    MEMBERS_CREDITED = MEMBERS_TOP = ()


__all__ = ["ADDON"]


class ADDON(object):
    """Class storing general add-on properties
    Property names need to be all-uppercase with no leading underscores
    """

    NAME = "Visual Feedback for Reviews"
    MODULE = "review_feedback"
    ID = "1749604199"
    VERSION = __version__
    LICENSE = "GNU AGPLv3"
    AUTHORS = (
        {
            "name": "Aristotelis P. (Glutanimate)",
            "years": "2017-2020",
            "contact": "Glutanimate",
        },
    )
    AUTHOR_MAIL = "ankiglutanimate [αt] gmail . com"
    LIBRARIES = ()
    CONTRIBUTORS = ()
    SPONSORS = ()
    MEMBERS_CREDITED = MEMBERS_CREDITED
    MEMBERS_TOP = MEMBERS_TOP
    LINKS = {
        "patreon": "https://www.patreon.com/glutanimate",
        "bepatron": "https://www.patreon.com/bePatron?u=7522179",
        "coffee": "http://ko-fi.com/glutanimate",
        "description": "https://ankiweb.net/shared/info/{}".format(ID),
        "rate": "https://ankiweb.net/shared/review/{}".format(ID),
        "twitter": "https://twitter.com/glutanimate",
        "youtube": "https://www.youtube.com/c/glutanimate",
        "help": "",
    }
